File: web/src/api/entityMentions.ts
import { httpContract, type HttpContractSuccessResponse } from './httpContract';
import type {
  EntityMentionDocumentsPage,
  EntityMentionGroup,
  EntityMentionOccurrencesPage,
  EntityMentionSelector,
  EntityMentionSurface,
  EntityMentionTypeTotal,
  EntityMentionsPreview,
} from './types';

export interface EntityMentionsOptions {
  signal?: AbortSignal;
  headers?: HeadersInit;
}

export interface EntityMentionsPreviewInput {
  sheetId: string;
  columnId: string;
  search?: string;
  type?: string;
  limit?: number;
  offset?: number;
}

export interface EntityMentionDocumentsInput {
  sheetId: string;
  columnId: string;
  type: string;
  fingerprint?: string;
  text?: string;
  limit?: number;
  offset?: number;
}

export interface EntityMentionOccurrencesInput {
  sheetId: string;
  rowId: string;
  columnId: string;
  type: string;
  fingerprint?: string;
  text?: string;
  limit?: number;
  offset?: number;
  snippetRadius?: number;
}

type EntityMentionsPreviewWire =
  HttpContractSuccessResponse<'tenant.entity_mentions_preview.post'>;
type EntityMentionDocumentsWire =
  HttpContractSuccessResponse<'tenant.entity_mention_documents.post'>;
type EntityMentionOccurrencesWire =
  HttpContractSuccessResponse<'tenant.entity_mention_occurrences.post'>;
type EntityMentionSelectorWire =
  | EntityMentionsPreviewWire['items'][number]['selector']
  | EntityMentionDocumentsWire['selector']
  | EntityMentionOccurrencesWire['selector'];

export interface EntityMentionsApi {
  preview(
    projectId: string,
    input: EntityMentionsPreviewInput,
    options?: EntityMentionsOptions,
  ): Promise<EntityMentionsPreview>;
  documents(
    projectId: string,
    input: EntityMentionDocumentsInput,
    options?: EntityMentionsOptions,
  ): Promise<EntityMentionDocumentsPage>;
  occurrences(
    projectId: string,
    input: EntityMentionOccurrencesInput,
    options?: EntityMentionsOptions,
  ): Promise<EntityMentionOccurrencesPage>;
}

type ContractErrorFactory = (status: number, payload: unknown) => Error;

function entityMentionSelector(
  selector: EntityMentionSelectorWire,
): EntityMentionSelector {
  if (selector.kind === 'fingerprint') {
    return { kind: 'fingerprint', fingerprint: selector.fingerprint };
  }
  return { kind: 'text', text: selector.text };
}

function mapDocuments(
  wire: EntityMentionDocumentsWire,
): EntityMentionDocumentsPage {
  return {
    sheetId: wire.sheet_id.toString(),
    columnId: wire.column.id.toString(),
    type: wire.type,
    selector: entityMentionSelector(wire.selector),
    totals: {
      mentions: wire.totals.mentions,
      documents: wire.totals.documents,
    },
    documents: wire.documents.map((document) => ({
      rowId: document.row_id.toString(),
      title: document.title,
      occurrenceCount: document.occurrence_count,
    })),
    nextOffset: wire.next_offset,
  };
}

function mapOccurrences(
  wire: EntityMentionOccurrencesWire,
): EntityMentionOccurrencesPage {
  return {
    sheetId: wire.sheet_id.toString(),
    rowId: wire.row_id.toString(),
    columnId: wire.column.id.toString(),
    type: wire.type,
    selector: entityMentionSelector(wire.selector),
    textColumn: wire.text_column === null
      ? null
      : {
          id: wire.text_column.id.toString(),
          name: wire.text_column.name,
        },
    totals: { occurrences: wire.totals.occurrences },
    occurrences: wire.occurrences.map((occurrence) => ({
      occurrenceId: occurrence.occurrence_id,
      start: occurrence.start,
      end: occurrence.end,
      quote: occurrence.quote,
      snippet: {
        text: occurrence.snippet.text,
        markStart: occurrence.snippet.mark_start,
        markEnd: occurrence.snippet.mark_end,
        truncatedStart: occurrence.snippet.truncated_start,
        truncatedEnd: occurrence.snippet.truncated_end,
      },
    })),
    nextOffset: wire.next_offset,
    unpositioned: wire.unpositioned === null
      ? null
      : { reason: wire.unpositioned.reason, total: wire.unpositioned.total },
  };
}

function mapPreview(
  wire: EntityMentionsPreviewWire,
): EntityMentionsPreview {
  return {
    sheetId: wire.sheet_id.toString(),
    column: {
      id: wire.column.id.toString(),
      name: wire.column.name,
      semanticType: wire.column.semantic_type,
    },
    coverage: {
      targetRows: wire.coverage.target_rows,
      completedRows: wire.coverage.completed_rows,
      failedRows: wire.coverage.failed_rows,
      sheetRows: wire.coverage.sheet_rows,
      scopeKind: wire.coverage.scope_kind,
    },
    search: wire.search,
    type: wire.type,
    totalGroups: wire.total_groups,
    typeTotals: wire.type_totals.map((total): EntityMentionTypeTotal => ({
      type: total.type,
      totalGroups: total.total_groups,
    })),
    limit: wire.limit,
    offset: wire.offset,
    items: wire.items.map((item): EntityMentionGroup => ({
      type: item.type,
      selector: entityMentionSelector(item.selector),
      label: item.label,
      rowCount: item.row_count,
      mentionCount: item.mention_count,
      surfaceCount: item.surface_count,
      surfaces: item.surfaces.map((surface): EntityMentionSurface => ({
        text: surface.text,
        rowCount: surface.row_count,
        mentionCount: surface.mention_count,
      })),
    })),
  };
}

export function createEntityMentionsApi(
  errorFactory: ContractErrorFactory,
): EntityMentionsApi {
  return {
    preview(projectId, input, options = {}) {
      return httpContract(
        'tenant.entity_mentions_preview.post',
        {
          pathParams: { pid: projectId },
          query: {},
          body: {
            sheet_id: Number(input.sheetId),
            column_id: Number(input.columnId),
            ...(input.search?.trim() ? { search: input.search } : {}),
            ...(input.type?.trim() ? { type: input.type } : {}),
            ...(input.limit != null ? { limit: input.limit } : {}),
            ...(input.offset != null ? { offset: input.offset } : {}),
          },
          signal: options.signal,
          headers: options.headers,
          errorFactory,
        },
        mapPreview,
      );
    },

    documents(projectId, input, options = {}) {
      return httpContract(
        'tenant.entity_mention_documents.post',
        {
          pathParams: { pid: projectId },
          query: {},
          body: {
            sheet_id: Number(input.sheetId),
            column_id: Number(input.columnId),
            type: input.type,
            ...(input.fingerprint !== undefined ? { fingerprint: input.fingerprint } : {}),
            ...(input.text !== undefined ? { text: input.text } : {}),
            ...(input.limit != null ? { limit: input.limit } : {}),
            ...(input.offset != null ? { offset: input.offset } : {}),
          },
          signal: options.signal,
          headers: options.headers,
          errorFactory,
        },
        mapDocuments,
      );
    },

    occurrences(projectId, input, options = {}) {
      return httpContract(
        'tenant.entity_mention_occurrences.post',
        {
          pathParams: { pid: projectId },
          query: {},
          body: {
            sheet_id: Number(input.sheetId),
            row_id: Number(input.rowId),
            column_id: Number(input.columnId),
            type: input.type,
            ...(input.fingerprint !== undefined ? { fingerprint: input.fingerprint } : {}),
            ...(input.text !== undefined ? { text: input.text } : {}),
            ...(input.limit != null ? { limit: input.limit } : {}),
            ...(input.offset != null ? { offset: input.offset } : {}),
            ...(input.snippetRadius != null ? { snippet_radius: input.snippetRadius } : {}),
          },
          signal: options.signal,
          headers: options.headers,
          errorFactory,
        },
        mapOccurrences,
      );
    },
  };
}

File: src/frisket/contracts/http/entity_mentions.py
"""Compatibility contracts for the entity-mention read routes.

The preview service owns validation for object payloads so its bare v1 action
errors remain stable. Requests therefore preserve raw values and ignore
extensions, while FastAPI continues to reject non-object JSON with 422.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import ConfigDict, Field, JsonValue

from frisket.contracts.http.models import CoerciveRequest, WireModel


class _CompatibleRequest(CoerciveRequest):
    pass


class EntityMentionsPreviewRequest(_CompatibleRequest):
    sheet_id: Any = None
    column_id: Any = None
    search: Any = None
    type: Any = None
    limit: Any = None
    offset: Any = None


class EntityMentionDocumentsRequest(_CompatibleRequest):
    sheet_id: Any = None
    column_id: Any = None
    type: Any = None
    fingerprint: Any = None
    text: Any = None
    limit: Any = None
    offset: Any = None


class EntityMentionOccurrencesRequest(_CompatibleRequest):
    sheet_id: Any = None
    row_id: Any = None
    column_id: Any = None
    type: Any = None
    fingerprint: Any = None
    text: Any = None
    limit: Any = None
    offset: Any = None
    snippet_radius: Any = None


class _OpenEntityMentionResponse(WireModel):
    """Strict mention core with lossless producer-owned JSON additions."""

    model_config = ConfigDict(extra="allow", strict=True)

    __pydantic_extra__: dict[str, JsonValue] = Field(init=False)


class EntityMentionColumn(_OpenEntityMentionResponse):
    id: int
    name: str
    semantic_type: str | None


class EntityMentionFingerprintSelector(_OpenEntityMentionResponse):
    kind: Literal["fingerprint"]
    fingerprint: str


class EntityMentionTextSelector(_OpenEntityMentionResponse):
    kind: Literal["text"]
    text: str


EntityMentionSelector = EntityMentionFingerprintSelector | EntityMentionTextSelector


class EntityMentionsCoverage(_OpenEntityMentionResponse):
    target_rows: int | None
    completed_rows: int | None
    failed_rows: int | None
    sheet_rows: int | None
    scope_kind: Literal["all_rows", "exact_membership"] | None


class EntityMentionSurface(_OpenEntityMentionResponse):
    text: str
    row_count: int
    mention_count: int


class EntityMentionGroup(_OpenEntityMentionResponse):
    type: str
    selector: EntityMentionSelector
    label: str
    row_count: int
    mention_count: int
    surface_count: int
    surfaces: list[EntityMentionSurface]


class EntityMentionTypeTotal(_OpenEntityMentionResponse):
    type: str
    total_groups: int


class EntityMentionsPreviewResponse(_OpenEntityMentionResponse):
    """Required mention inventory core plus additive producer JSON fields."""

    schema_version: Literal["entity-mentions-preview.v2"]
    sheet_id: int
    column: EntityMentionColumn
    coverage: EntityMentionsCoverage
    search: str | None
    type: str | None
    total_groups: int
    type_totals: list[EntityMentionTypeTotal]
    limit: int
    offset: int
    items: list[EntityMentionGroup]


class EntityMentionTotals(_OpenEntityMentionResponse):
    documents: int
    mentions: int


class EntityMentionDocument(_OpenEntityMentionResponse):
    row_id: int
    title: str | None
    occurrence_count: int


class EntityMentionDocumentsResponse(_OpenEntityMentionResponse):
    """Required document page core plus additive producer JSON fields."""

    schema_version: Literal["entity-mention-detail.v1"]
    sheet_id: int
    column: EntityMentionColumn
    type: str
    selector: EntityMentionSelector
    totals: EntityMentionTotals
    limit: int
    offset: int
    documents: list[EntityMentionDocument]
    next_offset: int | None


class EntityMentionOccurrencesTotals(_OpenEntityMentionResponse):
    occurrences: int


class EntityMentionTextColumn(_OpenEntityMentionResponse):
    id: int
    name: str | None


class EntityMentionSnippet(_OpenEntityMentionResponse):
    text: str
    mark_start: int
    mark_end: int
    truncated_start: bool
    truncated_end: bool


class EntityMentionOccurrence(_OpenEntityMentionResponse):
    occurrence_id: str
    start: int
    end: int
    quote: str
    snippet: EntityMentionSnippet


class EntityMentionUnpositioned(_OpenEntityMentionResponse):
    reason: Literal[
        "content_hash_mismatch", "unsupported_offset_unit", "invalid_geometry"
    ]
    total: int


class EntityMentionOccurrencesResponse(_OpenEntityMentionResponse):
    """Required occurrence page core plus additive producer JSON fields."""

    schema_version: Literal["entity-mention-occurrences.v1"]
    sheet_id: int
    row_id: int
    column: EntityMentionColumn
    type: str
    selector: EntityMentionSelector
    snippet_radius: int
    limit: int
    offset: int
    text_column: EntityMentionTextColumn | None
    totals: EntityMentionOccurrencesTotals
    occurrences: list[EntityMentionOccurrence]
    next_offset: int | None
    unpositioned: EntityMentionUnpositioned | None


__all__ = [
    "EntityMentionDocumentsRequest",
    "EntityMentionDocumentsResponse",
    "EntityMentionOccurrencesRequest",
    "EntityMentionOccurrencesResponse",
    "EntityMentionsPreviewRequest",
    "EntityMentionsPreviewResponse",
]
